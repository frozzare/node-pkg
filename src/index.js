const fs = require('fs');
const path = require('path');
const { build: esbuild } = require('esbuild');

export const build = async (entryPoint, config = {}) => {
  const out = await esbuild({
    ...config,
    entryPoints: [entryPoint],
    bundle: true,
    write: false,
  });

  const file = out.outputFiles[0];
  let text = file.text;

  if (config?.format === 'cjs') {
    const reg = /module.exports = \__toCommonJS\((\w+)\);/;
    const match = text.match(reg);
    const name = match[1].replace('_exports', '_default');

    text = text.replace(match[0], '') + `\nmodule.exports = ${name}`;
  }

  return !config.write ? text : fs.writeFileSync(file.path, text);
};
