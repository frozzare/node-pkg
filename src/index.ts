import fs from 'fs';
import { build as esbuild, BuildOptions } from 'esbuild';

export const build = async (
  entryPoint: string,
  config: Omit<BuildOptions, 'bundle' | 'entryPoints'> = {}
) => {
  const out = await esbuild({
    ...config,
    entryPoints: [entryPoint],
    bundle: true,
    write: false,
  });

  const file = out.outputFiles[0];
  let text = file.text;

  if (config.format === 'cjs') {
    const reg = /module.exports = \__toCommonJS\((\w+)\);/;
    const match = text.match(reg);
    const name = match[1].replace('_exports', '_default');

    text = text.replace(match[0], '') + `\nmodule.exports = ${name}`;
  }

  return !config.write ? text : fs.writeFileSync(file.path, text);
};
