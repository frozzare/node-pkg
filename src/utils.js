export const omit = (keys, obj) =>
  (Array.isArray(keys) ? keys : [keys]).reduce((a, e) => {
    const { [e]: omit, ...rest } = a;
    return rest;
  }, obj);
