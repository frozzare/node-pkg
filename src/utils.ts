export const omit = (keys, obj) => {
  if (!keys.length) return obj;
  const { [keys.pop()]: omitted, ...rest } = obj;
  return omit(keys, rest);
};
