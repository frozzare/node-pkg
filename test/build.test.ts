import { build } from '../src';

describe('build', () => {
  it('should build esm', async () => {
    const output = await build(`${__dirname}/fixtures/class.ts`, {
      write: false,
    });

    expect(output).toMatchSnapshot();
  });

  it('should build cjs', async () => {
    const output = await build(`${__dirname}/fixtures/class.ts`, {
      write: false,
      format: 'cjs',
    });

    expect(output).toMatchSnapshot();
    expect(output).toContain('module.exports = class_default');
  });

  it('should work with entryPoint as property', async () => {
    const output = await build({
      entry: `${__dirname}/fixtures/class.ts`,
      write: false,
    });

    expect(output).toMatchSnapshot();
  });
});
