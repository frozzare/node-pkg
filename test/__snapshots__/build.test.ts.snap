// Vitest Snapshot v1

exports[`build > should build cjs 1`] = `
"\\"use strict\\";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === \\"object\\" || typeof from === \\"function\\") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, \\"__esModule\\", { value: true }), mod);

// test/fixtures/class.ts
var class_exports = {};
__export(class_exports, {
  default: () => class_default
});

var Test = class {
};
var class_default = Test;
module.exports = class_default;"
`;

exports[`build > should build cjs with minify 1`] = `
"\\"use strict\\";var e=Object.defineProperty;var p=Object.getOwnPropertyDescriptor;var r=Object.getOwnPropertyNames;var s=Object.prototype.hasOwnProperty;var t=(l,a)=>{for(var d in a)e(l,d,{get:a[d],enumerable:!0})},u=(l,a,d,o)=>{if(a&&typeof a==\\"object\\"||typeof a==\\"function\\")for(let c of r(a))!s.call(l,c)&&c!==d&&e(l,c,{get:()=>a[c],enumerable:!(o=p(a,c))||o.enumerable});return l};var x=l=>u(e({},\\"__esModule\\",{value:!0}),l);var g={};t(g,{default:()=>b});var f=class{},b=f;
module.exports=b;"
`;

exports[`build > should build esm 1`] = `
"\\"use strict\\";
(() => {
  // test/fixtures/class.ts
  var Test = class {
  };
  var class_default = Test;
})();
"
`;

exports[`build > should work with entry as property 1`] = `
"\\"use strict\\";
(() => {
  // test/fixtures/class.ts
  var Test = class {
  };
  var class_default = Test;
})();
"
`;
